tests/test_flow/test_persistence.py

```diff
@@ -902,6 +902,73 @@ def wrapped_f(x, y):  # noqa: F811
     assert call_counter.times_called() == 1
 
 
+@pytest.mark.parametrize("versioning_mode", ["manual", "assist"])
+def test_versioning_class_changes(builder, make_counter, versioning_mode):
+    call_counter = make_counter()
+
+    # PicklableProtocol cannot pickle IntWrapper class because it is
+    # defined inside the test function. We instead use this custom
+    # protocol that wraps the PicklableProtocol to be able to serialize
+    # IntWrapper.
+    class IntWrapperProtocol(bn.protocols.PicklableProtocol):
+        def write(self, wrapper, path):
+            assert isinstance(wrapper, IntWrapper)
+            super().write(wrapper.value, path)
+
+        def read(self, path):
+            return IntWrapper(super().read(path))
+
+    class IntWrapper:
+        def __init__(self, value):
+            self.value = value
+
+        def __str__(self):
+            return f"IntWrapper: value={self.value}"
+
+    builder.set("core__versioning_mode", versioning_mode)
+
+    builder.assign("x", 2)
+    builder.assign("y", 3)
+
+    @builder
+    @IntWrapperProtocol()
+    @bn.version(1)
+    def wrapped_f(x, y):
+        return IntWrapper(x + y)
+
+    @builder
+    def f(wrapped_f):
+        call_counter.mark()
+        return wrapped_f.value
+
+    assert builder.build().get("f") == 5
+    assert builder.build().get("f") == 5
+    assert call_counter.times_called() == 1
+
+    builder.delete("wrapped_f")
+
+    # Make a change in the __str__ method.
+    class IntWrapper:  # noqa: F811
+        def __init__(self, value):
+            self.value = value
+
+        def __str__(self):
+            return f"value={self.value}"
+
+    @builder  # noqa: F811
+    @IntWrapperProtocol()
+    @bn.version(2)
```

Comment:
Suppose we used the same version here (`@version(1)`), and we had `versioning_mode == "assist"`. In that case we'd want this to throw an exception, right? Will that happen?